File: Python3/401 - 600/525 Contiguous Array.py
# 932ms, 97.32%; 21.8MB, 16.67%
class Solution:
    def findMaxLength(self, nums: list) -> int:
        length = len(nums)
        res, record, index = 0, 0, {0:[-1]}
        for i, n in enumerate(nums):
            record += 1 if n == 0 else -1
            if record in index:
                tmp = index[record]
                if len(tmp) == 2:
                    tmp[1] = i
                else:
                    tmp.append(i)
            else:
                index[record] = [i]
        if record == 0: return length
        for i in index:
            res = max(res, index[i][-1] - index[i][0])
        return res

File: 401 - 600/525. Contiguous Array.cc
// 128ms, 54.14%; 26.7MB, 16.67%
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int findMaxLength(const vector<int> &nums) {
        int length = nums.size();
        int res = 0, record = 0;
        unordered_map<int, vector<int>> index;
        index[0].push_back(-1);
        for (int i = 0; i < length; ++i) {
            record += (nums[i] == 0 ? 1 : -1);
            auto &it = index[record];
            if (it.size() == 2) {
                it[1] = i;
            } else {
                it.push_back(i);
            }
        }
        if (record == 0) return length;
        for (const auto &idx : index) {
            const auto &v = idx.second;
            res = max(res, v.back() - v.front());
        }
        return res;
    }
};
